internal/impl/qdrant/output.go

```diff
@@ -0,0 +1,237 @@
+	if err != nil {
+		return nil, err
+	}
+
+	vectorMapping, err := conf.FieldBloblang(qoFieldVectorMapping)
+	if err != nil {
+		return nil, err
+	}
+
+	payloadMapping, err := conf.FieldBloblang(qoFieldPayloadMapping)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := newQdrantClient(host, apiToken, enabled, config, mgr.Logger())
+	if err != nil {
+		return nil, err
+	}
+
+	w := outputWriter{
+		client: client,
+
+		collectionName: collectionName,
+		id:             id,
+		vectorMapping:  vectorMapping,
+		payloadMapping: payloadMapping,
+	}
+	return &w, nil
+}
+
+func (w *outputWriter) Connect(ctx context.Context) error {
+	return w.client.Connect(ctx)
+}
+
+func (w *outputWriter) WriteBatch(ctx context.Context, batch service.MessageBatch) (err error) {
+	batches, err := w.batchPointsByCollection(batch)
+	if err != nil {
+		return err
+	}
+	for cn, batch := range batches {
+		if err := w.client.Upsert(ctx, cn, batch); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *outputWriter) batchPointsByCollection(batch service.MessageBatch) (map[string][]*pb.PointStruct, error) {
+	cnExec := batch.InterpolationExecutor(w.collectionName)
+	idExec := batch.BloblangExecutor(w.id)
+	vectorExec := batch.BloblangExecutor(w.vectorMapping)
+	payloadExec := batch.BloblangExecutor(w.payloadMapping)
+	batches := map[string][]*pb.PointStruct{}
+	for i := 0; i < len(batch); i++ {
+		collectionName, err := cnExec.TryString(i)
+		if err != nil {
+			return nil, fmt.Errorf("%s interpolation error: %w", qoFieldCollectionName, err)
+		}
+		rawID, err := idExec.QueryValue(i)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute %s: %w", qoFieldID, err)
+		}
+
+		id, err := newPointID(rawID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to coerce point ID type: %w", err)
+		}
+
+		rawVec, err := vectorExec.Query(i)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute %s: %w", qoFieldVectorMapping, err)
+		}
+		if rawVec == nil {
+			continue
+		}
+		maybeVec, err := rawVec.AsStructured()
+		if err != nil {
+			return nil, fmt.Errorf("%s extraction failed: %w", qoFieldVectorMapping, err)
+		}
+		vec, err := newVectors(maybeVec)
+		if err != nil {
+			return nil, fmt.Errorf("unable to coerce vector output type: %w", err)
+		}
+
+		rawMeta, err := payloadExec.Query(i)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute %s: %w", qoFieldPayloadMapping, err)
+		}
+
+		maybePayload, err := rawMeta.AsStructured()
+		if err != nil {
+			return nil, fmt.Errorf("%s extraction failed: %w", qoFieldPayloadMapping, err)
+		}
+		payload, err := newValueMap(maybePayload)
+		if err != nil {
+			return nil, fmt.Errorf("unable to coerce payload output type: %w", err)
+		}
+
+		batches[collectionName] = append(batches[collectionName], &pb.PointStruct{
+			Id:      id,
+			Vectors: vec,
+			Payload: payload,
+		})
+	}
+	return batches, nil
+}
+
+func (w *outputWriter) Close(ctx context.Context) error {
+	if w.client != nil {
```

Comment:
The nil check here is probably superfluous unless `w.client.Connect(ctx)` can make it nil. Do we need any locking to make sure `Connect()` has executed successfully before attempting to write / close?